### PR TITLE
Move findTargetingStackObject to GameImpl, so saved temp data can be cleared

### LINKED ISSUE #13523: Ward not properly triggering when re-casting aura
Setup:
[[Adrix and Nev]] on the battlefield
[[Kenrith's Transformation]], [[Beast Within]] and [[Bala Ged Recovery]] in hand

Cast Kenrith on Adrivx it asks to pay ward {2} properly.
Destroy Kenrith with Beast Within, then bring it back with Bala Ged and recast it targeting the same creature
No ward triggers.

### PR DESCRIPTION
**Pull request by xenohedron**
This resolves #13523.

The `findTargetingStackObject` method includes a check for whether the trigger/watcher has already found that targeting object for that ability, in order to follow the rules correctly for multiple triggers/targets. Unfortunately, that data was persisting longer than needed and causing spurious failures when recasting a spell with the same target. It only needs to last as long as short living lki to function.

I determined that the simplest way to clear it was to move it from game state values (accessed via a CardUtil method) to a dedicated field in GameImpl (now accessed through a Game method). This allows it to be cleared at the appropriate time.

Test cases were added to WardTest to verify that it solves the reported problem. BecomesTargetTriggerTest verifies no regressions from previously reported issues.

Hopefully this finally covers all the edge cases with targeting triggers.

### CODE REVIEW

File: Mage/src/main/java/mage/game/GameImpl.java [hunk 1]
```diff
@@ -121,6 +121,8 @@ public abstract class GameImpl implements Game {
     protected Map<UUID, Map<Integer, MageObject>> lkiExtended = new HashMap<>();
     // Used to check if an object was moved by the current effect in resolution (so Wrath like effect can be handled correctly)
     protected Map<Zone, Set<UUID>> lkiShortLiving = new EnumMap<>(Zone.class);
+    // For checking "becomes the target" triggers accurately. Cleared on short living LKI reset
+    protected Map<String, Map<UUID, Set<UUID>>> targetedMap = new HashMap<>();
```
Comment (JayDi85) [inline Mage/src/main/java/mage/game/GameImpl.java:125]: 1. Check that table and make sure it’s really that you want (about short living logic lifecycle):
IMG_1534
2. If true then rename to lkiTargetedMap or like that
3. Create new issue about rework and remove that data from game object to game state or watcher. Or it must be rethink from scratch with new approach. It’s smells more harder now (writing on reading, diff results in same calls, data store outside from watcher and even game state, etc)

Comment (JayDi85) [reply Mage/src/main/java/mage/game/GameImpl.java:125]: 4. Also I recommend to check #8802 with current PR (is it works fine after one and more cancels). In normal use case a “cancel” by a user will execute rollback to original game state. But rollback restore only GameState data, so all GameImpl data will keep old info.

Comment (xenohedron) [reply Mage/src/main/java/mage/game/GameImpl.java:125]: It's correct - the info is only used for handling triggers in the moment, and not needed after that.

But it's not the same as "last known information" so that would be misleading to name it that. It just needs the same timing. The doc comments explain the usage.

Maybe it's better to rename reset LKI methods to describe when they reset. LKI reset on empty stack, short living LKI reset on stack object resolve. As this usage demonstrates, data other than "last known information" can have those lifecycle timing needs.

I do not believe it is appropriate to use a watcher for this data. It is a short living data that other watchers need to reference. Watchers are primarily designed for lifecycle that resets on turn end.

It does seem like it would be better in GameState. However, the LKI and shortLivingLKI is already in GameImpl instead of GameState. Do you know why that is? Because the data is so transient, I think it can be okay. If you think it's a problem, you can make an issue to research move all that info from GameImpl to GameState:

Screenshot_20250610_012216

Comment (xenohedron) [reply Mage/src/main/java/mage/game/GameImpl.java:125]: > 4. Also I recommend to check ["Becomes the target" abilities are triggered by selecting and undoing #8802](https://github.com/magefree/mage/issues/8802) with current PR (is it works fine after one and more cancels). In normal use case a “cancel” by a user will execute rollback to original game state. But rollback restore only GameState data, so all GameImpl data will keep old info.

Hm, good point. I think the root cause of that bug was completely eliminated by the changed approach, so it should be fine, but it does seem worth testing. If it's broken, then it'll have to be moved to GameState

Comment (JayDi85) [reply Mage/src/main/java/mage/game/GameImpl.java:125]: > Watchers are primarily designed for lifecycle that resets on turn end.

Nope, watchers to store card/ability/effect specific data. It’s not require to be clean on end turn. Moreover many watchers use multiple events logic — one for data collection, another one for clean. In your use case it can be something like trigger/activate event for cleanup (before targeting starting), target event for data collection. It’s only example.

> However, the LKI and shortLivingLKI is already in GameImpl instead of GameState. Do you know why that is?

Because there weren’t GameState in old days and all data stored in Game object. Most of the changeable fields moved to GameState, but some still waiting for refactor: https://github.com/magefree/mage/issues/2072#issuecomment-644327403

Comment (xenohedron) [reply Mage/src/main/java/mage/game/GameImpl.java:125]: > 4. Also I recommend to check ["Becomes the target" abilities are triggered by selecting and undoing #8802](https://github.com/magefree/mage/issues/8802) with current PR (is it works fine after one and more cancels). In normal use case a “cancel” by a user will execute rollback to original game state. But rollback restore only GameState data, so all GameImpl data will keep old info.

I tested this again with the new approach from this PR, and could not reproduce the bug.
